### PR TITLE
Add RC option for landing flare: idle throttle, tilts up, pitch= LAND_PITCH_CD

### PR DESCRIPTION
**Pull request by Hwurzburg**
allows emergency belly landing of all tilt rotors.... allows belly landing tailsitters and tilt rotors in fixed wing modes if battery too low for VTOL landing (allowing extra endurance by 2-3X) and  first step toward adding this also in autolanding final flare

only for manual throttle fixed wing modes not in VTOL transitions...

### CODE REVIEW

**Comment by tridge** [comment]
looks good!
how is testing going?

**Comment by Hwurzburg** [comment]
SITL worked fine with TVBS in manual, fbwa, and stabilize ...bail outs were very smooth...will try Convergence tomorrow....and then a Glider.....already  had tested that the qualifiers prevent use in other FW modes when it was a option_bit using throttle and elevator...

**Comment by Hwurzburg** [comment]
surprise, surprise, full up tilt is opposite for tailsitters and regular tilt rotors....had to add a line for that....dont expect any surprises from conventional planes...

**Comment by Hwurzburg** [comment]
and bicopter quadplane is different yet again...fixed...now tested on Convergence, bicopter,tvbs, and normal plane...think its good to go...

**Comment by Hwurzburg** [comment]
a few final tweaks....use THR_MIN instead of 0 to keep ICE running....and block throttle change unless armed to avoid bench runs at THR_MIN...will test fly on my new TVBS today

File: ArduPlane/servos.cpp [hunk 1]
```diff
@@ -902,6 +903,26 @@ void Plane::servos_output(void)
...
+         SRV_Channels::set_output_scaled(SRV_Channel::k_motor_tilt, tilt);
+         SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorLeft, tilt);
+         SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorRight, tilt);
+         float throttle_min = aparm.throttle_min.get(); //allows ICE to run if used
+             if (arming.is_armed()) {  //prevent running motors if unarmed
+                 SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, throttle_min);
+                 SRV_Channels::set_output_scaled(SRV_Channel::k_throttleLeft, throttle_min);
+                 SRV_Channels::set_output_scaled(SRV_Channel::k_throttleRight, throttle_min);
+             }
+     }
```
Comment (IamPete1) [inline ArduPlane/servos.cpp:924]: would be nice to have this in its own little function. like `servos_twin_engine_mix()`

File: ArduPlane/servos.cpp [hunk 2]
```diff
@@ -902,6 +903,26 @@ void Plane::servos_output(void)
     quadplane.tailsitter_output();
     quadplane.tiltrotor_bicopter();
 
+     // pull up tilts and set throttle_min if LANDING_FLARE RCx_OPTION switch is set and in FW mode, manual throttle
+     if (!quadplane.in_transition() && !control_mode->is_vtol_mode() && !auto_throttle_mode && plane.channel_throttle->norm_input() < -0.95 && plane.flare_switch_active) {
+         int32_t tilt = -4500;
```
Comment (IamPete1) [inline ArduPlane/servos.cpp:908]: You can use the #define 
```suggestion
         int32_t tilt = -SERVO_MAX;
```

File: ArduPlane/servos.cpp [hunk 3]
```diff
@@ -902,6 +903,26 @@ void Plane::servos_output(void)
...
+         if (quadplane.tilt.tilt_type == QuadPlane::TILT_TYPE_BICOPTER) {
+                 tilt = 0;
+         }
+         if (quadplane.is_tailsitter()) {
+                 tilt = 4500;
```
Comment (IamPete1) [inline ArduPlane/servos.cpp:913]:
```suggestion
                 tilt = SERVO_MAX;
```

File: ArduPlane/servos.cpp [hunk 4]
```diff
@@ -902,6 +903,26 @@ void Plane::servos_output(void)
...
+         }
+         SRV_Channels::set_output_scaled(SRV_Channel::k_motor_tilt, tilt);
+         SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorLeft, tilt);
+         SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorRight, tilt);
+         float throttle_min = aparm.throttle_min.get(); //allows ICE to run if used
+             if (arming.is_armed()) {  //prevent running motors if unarmed
```
Comment (IamPete1) [inline ArduPlane/servos.cpp:919]: This if is unnecessarily indented

File: ArduPlane/servos.cpp [hunk 5]
```diff
@@ -881,6 +881,7 @@ void Plane::set_servos(void)
         SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, override_pct);
     }
 
+
```
Comment (IamPete1) [inline ArduPlane/servos.cpp:884]: accidental white-space change

Comment (Hwurzburg) [reply ArduPlane/servos.cpp]: fixed

File: ArduPlane/servos.cpp (hunk 4 above)
Comment (Hwurzburg) [reply ArduPlane/servos.cpp]: fixed in function

File: ArduPlane/servos.cpp (hunk 3 above)
Comment (Hwurzburg) [reply ArduPlane/servos.cpp]: done

File: ArduPlane/servos.cpp (hunk 2 above)
Comment (Hwurzburg) [reply ArduPlane/servos.cpp]: done

File: ArduPlane/servos.cpp (hunk 1 above)
Comment (Hwurzburg) [reply ArduPlane/servos.cpp]: done

File: ArduPlane/servos.cpp [hunk 6]
```diff
@@ -727,9 +727,35 @@ void Plane::servos_twin_engine_mix(void)
...
+        }
+        if (quadplane.is_tailsitter()) {
+            tilt = SERVO_MAX;
+        }
+        SRV_Channels::set_output_scaled(SRV_Channel::k_motor_tilt, tilt);
+        SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorLeft, tilt);
+        SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorRight, tilt);
+        float throttle_min = aparm.throttle_min.get(); //allows ICE to run if used
+        if (arming.is_armed()) {  //prevent running motors if unarmed
+            SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, throttle_min);
```
Comment (magicrub) [inline ArduPlane/servos.cpp:751]: why isn't this using AP_LANDING somehow?

Comment (Hwurzburg) [reply ArduPlane/servos.cpp:751]: Tridge wants to eventually add it to autoland, but thats a bit ambitious for me....I just wanted a belly land option for tailsitters and tiltrotors and Tridge suggested I expand it a bit....future enhancement to add to final flare in autolanding

**Comment by Naterater** [comment]
> a few final tweaks....use THR_MIN instead of 0 to keep ICE running....and block throttle change unless armed to avoid bench runs at THR_MIN...will test fly on my new TVBS today

Keep in mind we use reverse thrust in Fixed-wing mode to traverse the back sides of hills without speeding up. THR_MIN could be full reverse thrust for BLHeli ESC aircraft.

**Comment by Hwurzburg** [comment]
@Naterater Thanks....nice catch...fixed...and SITL'd...

File: ArduPlane/RC_Channel.cpp [hunk 7]
```diff
@@ -116,6 +116,20 @@ void RC_Channel_Plane::do_aux_function_soaring_3pos(AuxSwitchPos ch_flag)
...
 #endif
 }
+
+void RC_Channel_Plane::do_aux_function_flare(AuxSwitchPos ch_flag)
+{
+        switch(ch_flag) {
+        case AuxSwitchPos::HIGH:
+            plane.flare_switch_active = true;
+            break;
+        case AuxSwitchPos::MIDDLE:
```
Comment (tridge) [inline ArduPlane/RC_Channel.cpp:126]: MIDDLE is normally do nothing for 2 pos switch (add a break)

File: ArduPlane/servos.cpp [hunk 8]
```diff
@@ -727,9 +727,35 @@ void Plane::servos_twin_engine_mix(void)
...
-  Set the flight control servos based on the current calculated values
+  Set throttle,attitude(in Attitude.cpp), and tilt servos for forced flare by RCx_OPTION switch for landing in FW mode
+  For Fixed Wind modes with manual throttle control only. Forces tilts up and throttle to THR_MIN.
+  Throttle stick must be in idle deadzone. This allows non-momentary switch to be used and quick bailouts
+  for go-arounds. Also helps prevent propstrike after landing with switch release on ground.
+*/
+void Plane::force_flare(void)
+{
+    if (!quadplane.in_transition() && !control_mode->is_vtol_mode() && !auto_throttle_mode && plane.channel_throttle->in_trim_dz() && plane.flare_switch_active) {
+        int32_t tilt = -SERVO_MAX;
```
Comment (tridge) [inline ArduPlane/servos.cpp:739]: a comment explaining why the 3 cases for tilt would be good

File: ArduPlane/servos.cpp [hunk 9]
```diff
@@ -727,9 +727,35 @@ void Plane::servos_twin_engine_mix(void)
...
 /*
-  Set the flight control servos based on the current calculated values
+  Set throttle,attitude(in Attitude.cpp), and tilt servos for forced flare by RCx_OPTION switch for landing in FW mode
+  For Fixed Wind modes with manual throttle control only. Forces tilts up and throttle to THR_MIN.
+  Throttle stick must be in idle deadzone. This allows non-momentary switch to be used and quick bailouts
+  for go-arounds. Also helps prevent propstrike after landing with switch release on ground.
+*/
+void Plane::force_flare(void)
+{
+    if (!quadplane.in_transition() && !control_mode->is_vtol_mode() && !auto_throttle_mode && plane.channel_throttle->in_trim_dz() && plane.flare_switch_active) {
```
Comment (tridge) [inline ArduPlane/servos.cpp:738]: can remote plane. prefix

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduPlane/servos.cpp (hunk 9 above)
Comment (Hwurzburg) [reply ArduPlane/servos.cpp:738]: done

File: ArduPlane/servos.cpp (hunk 8 above)
Comment (Hwurzburg) [reply ArduPlane/servos.cpp:739]: done

File: ArduPlane/RC_Channel.cpp (hunk 7 above)
Comment (Hwurzburg) [reply ArduPlane/RC_Channel.cpp:126]: done

**Comment by pompecukor** [comment]
Hi,

First of all, I would like to say thank you to the author.
This can be helpful in some cases where you are guaranteed a crash otherwise. For instance an almost dead battery with still manage glide to land. Hence I really like it and am greatful. Especially if there will be some way to even automate it (some kind of battery failsafe).
For this reason I would like to add two items to consider that will not be obvious is SITL, but can slap you in the face in real life.

1. (the more important one) is CG in most common tilt-rotors are CG correctly with the motors/props fully tilted forward.This means as soon as you tilt them up while you are still gliding as a forward flight plane, you will instantly become tail heavy.
This means the there should be a warning not to attempt this "flare" prematurely. As the operator might be in for a surprise. low airspeed (landing) and tail heavy don't mix well as we know.

2. This one I am not so sure is applicable as I understandit depends on flightmode,  but... there is an alt assistance I believe in the Q parameters. There has to be a way for the system to dissable that when attempting to land in this way. One of the uses as described for this feature is to permit a landing when battery is too low for hover. That is not the time we would want to user to have to dig into the parameters (if they even remeber) to dissable q assist alt. Before they can start descending.

**Comment by Hwurzburg** [comment]
@pompecukor thanks for the input...

as to #1, I doubt that it will be a problem for most tilt-rotors....if it were, I think that they would be in stall at the back half of the transition as VTOL is phased out....in the SITL RF sims and in the flight tests on two TVBS, it was not an issue...but its worth a note in the wiki when this gets described (Assuming its merged) and I will make sure that occurs

#2...another corner case!...thanks...I have added a force disable to Qassistance when the switch is activated....if already in assist, it wont do anything, but if Qassist is not active yet, it will disable it for as long the flare switch is high....bailouts will still be possible, but assistance wont be active again until the switch is back low...

**Review by tridge — APPROVED** [review]

**Review by magicrub — APPROVED** [review]
